src/packages/documents/documents/components/input-document-picker-root/input-document-picker-root.element.ts

```diff
@@ -1,101 +1,268 @@
+				this._originManifests = originManifests;
+			},
+		);
 
-		this.observe(this.#documentPickerContext.selection, (selection) => (super.value = selection.join(',')));
-		this.observe(this.#documentPickerContext.selectedItems, (selectedItems) => (this._items = selectedItems));
+		this.observe(
+			umbExtensionsRegistry.byType('dynamicRootQueryStep'),
+			(queryStepManifests: Array<ManifestDynamicRootQueryStep>) => {
+				this._queryStepManifests = queryStepManifests;
+			},
+		);
 	}
 
-	protected getFormElement() {
-		return undefined;
+	connectedCallback(): void {
+		super.connectedCallback();
+
+		this.#updateDynamicRootOrigin(this.data);
+		this.#updateDynamicRootQuerySteps(this.data?.querySteps);
+	}
+
+	#sorter?: UmbSorterController<UmbTreePickerDynamicRootQueryStep>;
+
+	#initSorter() {
+		this.#sorter = new UmbSorterController<UmbTreePickerDynamicRootQueryStep>(this, {
+			getUniqueOfElement: (element) => {
+				return element.id;
+			},
+			getUniqueOfModel: (modelEntry) => {
+				return modelEntry.unique;
+			},
+			identifier: 'Umb.SorterIdentifier.InputDocumentPickerRoot',
+			itemSelector: 'uui-ref-node',
+			containerSelector: '#query-steps',
+			onChange: ({ model }) => {
+				if (this.data && this.data.querySteps) {
+					//const steps = [...this.data.querySteps];
+					const querySteps = model; //.map((index) => steps[parseInt(index)]);
+					this.#updateDynamicRootQuerySteps(querySteps);
+					this.dispatchEvent(new UmbChangeEvent());
+				}
+			},
+		});
+	}
+
+	#openDynamicRootOriginPicker() {
+		this.#openModal = this.#modalContext?.open(UMB_DYNAMIC_ROOT_ORIGIN_PICKER_MODAL, {});
+		this.#openModal?.onSubmit().then((data: UmbTreePickerDynamicRoot) => {
+			const existingData = { ...this.data };
+			existingData.originKey = undefined;
+			this.data = { ...existingData, ...data };
+			this.#updateDynamicRootOrigin(this.data);
+			this.dispatchEvent(new UmbChangeEvent());
+		});
+	}
+
+	#openDynamicRootQueryStepPicker() {
+		this.#openModal = this.#modalContext?.open(UMB_DYNAMIC_ROOT_QUERY_STEP_PICKER_MODAL, {});
+		this.#openModal?.onSubmit().then((step) => {
+			if (this.data) {
+				const querySteps = [...(this.data.querySteps ?? []), step];
+				this.#updateDynamicRootQuerySteps(querySteps);
+				this.dispatchEvent(new UmbChangeEvent());
+			}
+		});
+	}
+
+	#updateDynamicRootOrigin(data?: UmbTreePickerDynamicRoot) {
+		if (!data) return;
+		const origin = this._originManifests.find((item) => item.meta.originAlias === data.originAlias)?.meta;
+		this.#dynamicRootOrigin = {
+			label: origin?.label ?? data.originAlias,
+			icon: origin?.icon ?? 'icon-wand',
+			description: data.originKey,
+		};
+	}
+
+	#updateDynamicRootQuerySteps(querySteps?: Array<UmbTreePickerDynamicRootQueryStep>) {
+		if (!this.data) return;
+
+		if (!this.#sorter) {
+			// NOTE: The sorter controller can only be initialized when the `#query-steps` element is available. [LK]
+			this.#initSorter();
+		}
```

Comment:
Is it possible at all to render an empty #query-steps uui-ref-list before any items are created?

Reply:
I've refactored the markup, so that the initial `<uui-ref-list>` elements are available the DOM.